etna/transforms/missing_values/imputation.py

```diff
@@ -185,29 +216,44 @@ def _fill(self, df: pd.DataFrame) -> pd.DataFrame:
         :
             Filled Dataframe.
         """
-        self._fill_value = cast(Dict[str, float], self._fill_value)
-        self._nan_timestamps = cast(Dict[str, List[pd.Timestamp]], self._nan_timestamps)
-        segments = sorted(set(df.columns.get_level_values("segment")))
+        if self._nans_to_impute_mask is None:
+            raise ValueError("Transform is not fitted!")
 
         if self._strategy is ImputerMode.constant:
-            new_values = df.loc[:, pd.IndexSlice[:, self.in_column]].fillna(value=self.constant_value)
-            df.loc[:, pd.IndexSlice[:, self.in_column]] = new_values
+            df.fillna(value=self.constant_value, inplace=True)
         elif self._strategy is ImputerMode.forward_fill:
-            new_values = df.loc[:, pd.IndexSlice[:, self.in_column]].fillna(method="ffill")
-            df.loc[:, pd.IndexSlice[:, self.in_column]] = new_values
+            df.fillna(method="ffill", inplace=True)
         elif self._strategy is ImputerMode.mean:
-            for segment in segments:
-                df.loc[:, pd.IndexSlice[segment, self.in_column]].fillna(value=self._fill_value[segment], inplace=True)
+            self._mean_imputer.transform(df)
         elif self._strategy is ImputerMode.running_mean or self._strategy is ImputerMode.seasonal:
-            timestamp_to_index = {timestamp: i for i, timestamp in enumerate(df.index)}
-            for segment in segments:
-                history = self.seasonality * self.window if self.window != -1 else len(df)
-                for timestamp in self._nan_timestamps[segment]:
-                    i = timestamp_to_index[timestamp]
-                    indexes = np.arange(i - self.seasonality, i - self.seasonality - history, -self.seasonality)
-                    indexes = indexes[indexes >= 0]
-                    values = df.loc[df.index[indexes], pd.IndexSlice[segment, self.in_column]]
-                    df.loc[timestamp, pd.IndexSlice[segment, self.in_column]] = np.nanmean(values)
+            history = self.seasonality * self.window if self.window != -1 else len(df)
+            nan_mask = df.isna().values
+            nan_indexes = np.arange(len(df))[nan_mask.any(axis=1)]
```

Comment:
Do I understand correctly that it is set to `True` if for any segment we have a `NaN` here?

Reply:
Yes

Reply:
`nan_indexes` contains indexes with at least one NaN